appvia/kore
# Deploy default domains to clusters

## Summary

This adds the ability to automatically deploy child zones when a root zone is created.

**Which issue(s) this PR resolves**:

Contributes toward #1298 

## Changes

Will create a 'ServiceDeployment' for each supported cluster type (and it's related cloud DNS provider).

This will result in a "child" `domain` `service` created for each cluster that exists / is created with the name of the cluster.

## Examples

Automatic child domains per cluster can be configured with the new Configuration setting:

```
  configuration:
    deployChildZoneRules:
      zonePerCluster: true
```

A full example for a kore "root" domain is given below:
```
apiVersion: services.kore.appvia.io/v1
kind: Service
metadata:
  name: root-domain
  namespace: kore-admin
spec:
  cluster:
    group: clusters.compute.kore.appvia.io
    kind: Cluster
    name: kore
    namespace: kore-admin
    version: v1
  configuration:
    dnsProvider: Route53
    deployClusterDNS: false
    domain: lewis-new.labs.kore.appvia.io
    credentials:
      name: aws-appvia
      namespace: kore-admin
      type: EKSCredentials
    private: false
    deployChildZoneRules:
      zonePerCluster: true
  kind: domain
  plan: domain
```

## Testing

1. Apply the manifest above or create a new domain in the Config / DNS page in the UI:
    ```
    kore apply -f ./manifests/local/dnsservice.yam
    ```

2. Manually delegate the zone to allow the status to come good.

3. Create / check that any EKS clusters exist

4. Check that child new domains have been created and the status is good:
    ```
    $ kore -t lewis-local get services
    NAME                                         KIND                        PLAN                              STATUS     AGE
    lewis-local-dns-root-domain-route53          domain                      domain                            Success    38m
    lewis-local-new-root-domain-route53          domain                      domain                            Success    36m
    ```

## Follow on Issues

Will simply add this option as the default when a user adds "a root domain" to kore but this will wait for pr #1345 as the controls have been reworked.
